### PR TITLE
Implement SetTraceTag client API

### PR DESCRIPTION
**Pull request by B-Step62**
<details><summary>&#x1F6E0 DevTools &#x1F6E0</summary>
<p>

[Open in GitHub Codespaces](https://codespaces.new/B-Step62/mlflow/pull/11565?quickstart=1)

#### Install mlflow from this PR

```
pip install git+https://github.com/mlflow/mlflow.git@refs/pull/11565/merge
```

#### Checkout with GitHub CLI

```
gh pr checkout 11565
```

</p>
</details>

### What changes are proposed in this pull request?

Implement `MlflowClient().set_trace_tag(request_id, key, value)` API. A tag can be  either an ongoing trace or a logged trace. Proto definition is just copied from Databricks side. 

### How is this PR tested?

- [x] Existing unit/integration tests
- [x] New unit/integration tests
- [x] Manual tests

Tested in staging notebook with local client. Haven't tested with databricks backend (not sure how to do it correctly).

### Does this PR require documentation update?

- [x] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [x] No. You can skip the rest of this section.
- [ ] Yes. Give a description of this change to be included in the release notes for MLflow users.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/deployments`: MLflow Deployments client APIs, server, and third-party Deployments integrations
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [x] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: mlflow/entities/trace_info.py [hunk 1]
```diff
@@ -71,3 +74,35 @@ def from_proto(cls, proto):
...
+                INVALID_PARAMETER_VALUE,
+            )
+
+        if len(value) > MAX_CHARS_IN_TRACE_INFO_METADATA_AND_TAGS:
+            raise MlflowException(
+                f"A value for a trace tag exceeds the maximum allowed length of "
+                f"{MAX_CHARS_IN_TRACE_INFO_METADATA_AND_TAGS} characters. "
+                "Got: {len(value)} for the key '{key}'",
+                INVALID_PARAMETER_VALUE,
+            )
```
Comment (BenWilson2) [inline mlflow/entities/trace_info.py:108]: Do we need this validation and the checks if we're automatically truncating these?

**Review by BenWilson2 — APPROVED** [review]
overall, LGTM - just one suggestion on type validation casting to mimic the behavior of other Tagging APIs. Lets make it consistent if there isn't a compelling reason not to.

File: mlflow/entities/trace_info.py (hunk 1 above)
Comment (dbczumar) [inline mlflow/entities/trace_info.py:108]: @B-Step62 @BenWilson2 We shouldn't be doing any client-side validation of tag or metadata length. That should be up to the backend. Client-side *truncation* is okay.

**Review by dbczumar — CHANGES_REQUESTED** [review]
Blocking until https://github.com/mlflow/mlflow/pull/11565/files#r1546692080 is addressed. Thanks @B-Step62

File: mlflow/entities/trace_info.py [hunk 2]
```diff
@@ -71,3 +74,35 @@ def from_proto(cls, proto):
...
+        """
+        Validate tag key and value.
+        """
+        if not key or not isinstance(key, str):
+            raise MlflowException(
+                f"A key for a trace tag must be a non-empty string. Got: {key}",
+                INVALID_PARAMETER_VALUE,
+            )
+
+        if not isinstance(value, str):
```
[the remark this answers is not in the document]
Comment (B-Step62) [reply mlflow/entities/trace_info.py:89]: Got it, doing `str()` like run tag makes sense.

File: mlflow/entities/trace_info.py (hunk 1 above)
Comment (B-Step62) [reply mlflow/entities/trace_info.py:108]: > when setting tags for truncated input & output value.

Understood, do you mean `request_metadata` here?

File: mlflow/protos/service.proto [hunk 3]
```diff
@@ -1294,3 +1294,21 @@ message GetTraceInfo {
...
+  // All storage backends are guaranteed to support key values up to 250 bytes in size.
+  optional string key = 2;
+
+  // String value of the tag being logged. Maximum size depends on storage backend.
+  // All storage backends are guaranteed to support key values up to 250 bytes in size.
+  optional string value = 3;
+
+  message Response {
+  }
+}
```
Comment (harupy) [inline mlflow/protos/service.proto:1314]:
```suggestion
}

```

File: mlflow/tracking/client.py [hunk 4]
```diff
@@ -713,6 +714,43 @@ def end_span(
...
+        """
+        Set a tag on the trace with the given trace ID.
+
+        The trace can be ongoing one or the one that has already ended and recorded in the backend.
+        Below is an example of setting a tag on an ongoing trace. You can replace the ``request_id``
+        parameter to setting a tag on an already ended trace.
+
+        .. code-block:: python
+
+            from mlflow.tracking import MlflowClient
```
Comment (harupy) [inline mlflow/tracking/client.py:727]:
```suggestion
            from mlflow import MlflowClient
```

File: mlflow/tracking/client.py [hunk 5]
```diff
@@ -713,6 +714,43 @@ def end_span(
 
         span.end()
 
+    def set_trace_tag(self, request_id: str, key: str, value: str):
+        """
+        Set a tag on the trace with the given trace ID.
+
+        The trace can be ongoing one or the one that has already ended and recorded in the backend.
```
Comment (harupy) [inline mlflow/tracking/client.py:721]: Does ongoing here mean active?

File: tests/tracking/test_client.py [hunk 6]
```diff
@@ -328,6 +328,28 @@ def test_start_span_raise_error_when_parent_span_id_is_not_provided():
         )
 
 
+def test_set_trace_tag_on_ongoing_trace(mock_trace_client):
+    client = mlflow.tracking.MlflowClient()
+
+    root_span = client.start_trace(name="test")
+    request_id = root_span.request_id
+    client.set_trace_tag(request_id, "foo", "bar")
+    client.end_trace(request_id)
```
Comment (harupy) [inline tests/tracking/test_client.py:337]: If line 336 fails, would that leave a zombie (?) span and affect other tests?

File: mlflow/tracking/client.py (hunk 5 above)
Comment (B-Step62) [reply mlflow/tracking/client.py:721]: A trace that is started and running i.e. not yet ended.

Comment (harupy) [reply mlflow/tracking/client.py:721]: can we use `active` because we use `active` for runs?

Comment (B-Step62) [reply mlflow/tracking/client.py:721]: makes sense!

File: mlflow/entities/trace_info.py [hunk 7]
```diff
@@ -54,7 +54,7 @@ def to_proto(self):
         for key, value in self.tags.items():
             tag = ProtoTraceRequestMetadata()
             tag.key = key
-            tag.value = value
+            tag.value = str(value)
```
Comment (dbczumar) [inline mlflow/entities/trace_info.py:57]: IMO this conversion should be done upstream in the client layer. This is how it's done for run tags.

Comment (dbczumar) [inline mlflow/entities/trace_info.py:57]: IMO this conversion should be done upstream in the client layer. This is how it's done for run tags.



```suggestion
            tag.value = value
```

File: mlflow/tracking/client.py [hunk 8]
```diff
@@ -713,6 +714,41 @@ def end_span(
...
+            value: The string value of the tag. Must be shorter than 250 characters.
+        """
+        # Trying to set the tag on the active trace first
+        trace_manager = InMemoryTraceManager.get_instance()
+        try:
+            trace_manager.set_trace_tag(request_id, key, value)
+            return
+        except MlflowException as e:
+            if e.error_code != ErrorCode.Name(RESOURCE_DOES_NOT_EXIST):
+                raise
```
Comment (dbczumar) [inline mlflow/tracking/client.py:747]: This is a brilliant catch. Well done, @B-Step62 :D

File: mlflow/tracking/client.py [hunk 9]
```diff
@@ -713,6 +714,41 @@ def end_span(
 
         span.end()
 
+    def set_trace_tag(self, request_id: str, key: str, value: str):
```
Comment (dbczumar) [inline mlflow/tracking/client.py:717]: I think we should do casting for `value` here

**Review by dbczumar — APPROVED** [review]
LGTM once https://github.com/mlflow/mlflow/pull/11565/files#r1548669238 is addressed. Thanks so much, @B-Step62 !

File: mlflow/entities/trace_info.py (hunk 7 above)
Comment (B-Step62) [reply mlflow/entities/trace_info.py:57]: makes sense!

**Pull request merged**